server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -168,6 +171,7 @@ public String toString() {
             if (event.localNodeMaster() == false) {
                 onNoLongerMaster();
             }
+            nodeIsMaster.set(event.localNodeMaster());
```

Comment:
Perhaps there is an alternative pattern for doing this kind of thing?

Reply:
Nit: we can reduce the number of atomic set calls with a check of `event.localNodeMaster() != event.previousState().nodes().isLocalNodeElectedMaster()`.

Alternatively, we can register the metrics in this class since we are already accessing the fields like `desiredBalanceReconciler.unassignedShards` directly in `onNoLongerMater`. This way we can keep a volatile field for master check and no need to pass it down. Not necessarily better than what you have. Happy for you to decide.

Reply:
I've done a fairly significant refactor to move all the metrics and their state & logic to a dedicated class, it means we can unit test, get rid of all the Atomics, and make a nicer interface for setting whether we're master and getting and setting the metric values.